kmod/patch/kpatch-macros.h

```diff
@@ -39,6 +27,84 @@ struct kpatch_unload {
 #define KPATCH_IGNORE_FUNCTION(_fn) \
 	void *__kpatch_ignore_func_##_fn __section(.kpatch.ignore.functions) = _fn;
 
+
+/* Support for livepatch callback hooks */
+#if IS_ENABLED(CONFIG_LIVEPATCH)
+# if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+#  define HAS_LIVEPATCH_HOOKS
+# endif
+#endif
+
+#ifdef HAS_LIVEPATCH_HOOKS
+
+#include <linux/livepatch.h>
+
+typedef int (*kpatch_pre_patchcall_t)(struct klp_object *obj);
+typedef void (*kpatch_post_patchcall_t)(struct klp_object *obj);
+typedef void (*kpatch_pre_unpatchcall_t)(struct klp_object *obj);
+typedef void (*kpatch_post_unpatchcall_t)(struct klp_object *obj);
+
+struct kpatch_pre_patch {
+	kpatch_pre_patchcall_t fn;
+	char *objname; /* filled in by create-diff-object */
+};
+
+struct kpatch_post_patch {
+	kpatch_post_patchcall_t fn;
+	char *objname; /* filled in by create-diff-object */
+};
+
+struct kpatch_pre_unpatch {
+	kpatch_pre_unpatchcall_t fn;
+	char *objname; /* filled in by create-diff-object */
+};
+
+struct kpatch_post_unpatch {
+	kpatch_post_unpatchcall_t fn;
+	char *objname; /* filled in by create-diff-object */
+};
+
+
+#define KPATCH_PRE_PATCH_HOOK(_fn) \
+	static inline kpatch_pre_patchcall_t __pre_patchtest(void) { return _fn; } \
+	struct kpatch_pre_patch kpatch_pre_patch_data_##_fn __section(.kpatch.hooks.pre_patch) = { \
```

Comment:
I think the `##_fn` isn't necessary, and instead the struct can be made static.

Reply:
If `kpatch_{pre|post}_{|un}patch_data` becomes static, we probably need to make sure the compiler won't optimize it away.

I am currently experimenting with the same test patch as above and, it seems, there are no ".kpatch.hooks*" sections in the patched object file.

Reply:
Yes, `__used` helped.